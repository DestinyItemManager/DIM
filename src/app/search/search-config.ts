import { DestinyVersion } from '@destinyitemmanager/dim-api-types';
import { destinyVersionSelector } from 'app/accounts/selectors';
import { createSelector } from 'reselect';
import { canonicalFilterFormats, FilterDefinition, SuggestionsContext } from './filter-types';
import advancedFilters from './search-filters/advanced';
import d1Filters from './search-filters/d1-filters';
import dupeFilters from './search-filters/dupes';
import freeformFilters from './search-filters/freeform';
import itemInfosFilters from './search-filters/item-infos';
import knownValuesFilters from './search-filters/known-values';
import loadoutFilters from './search-filters/loadouts';
import simpleRangeFilters from './search-filters/range-numeric';
import overloadedRangeFilters from './search-filters/range-overload';
import simpleFilters from './search-filters/simple';
import socketFilters from './search-filters/sockets';
import statFilters from './search-filters/stats';
import locationFilters from './search-filters/stores';
import wishlistFilters from './search-filters/wishlist';
import { generateSuggestionsForFilter, suggestionsContextSelector } from './suggestions-generation';

const allFilters = [
  ...dupeFilters,
  ...($featureFlags.wishLists ? wishlistFilters : []),
  ...freeformFilters,
  ...itemInfosFilters,
  ...knownValuesFilters,
  ...d1Filters,
  ...loadoutFilters,
  ...simpleRangeFilters,
  ...overloadedRangeFilters,
  ...simpleFilters,
  ...socketFilters,
  ...statFilters,
  ...locationFilters,
  ...advancedFilters,
];

export const searchConfigSelector = createSelector(
  destinyVersionSelector,
  suggestionsContextSelector,
  buildSearchConfig
);

//
// SearchConfig
//

export interface SearchConfig {
  allFilters: FilterDefinition[];
  /* `is:keyword` filters */
  isFilters: Record<string, FilterDefinition>;
  /* `keyword:value` filters */
  kvFilters: Record<string, FilterDefinition>;
  suggestions: string[];
}

/** Builds an object that describes the available search keywords and filter definitions. */
export function buildSearchConfig(
  destinyVersion: DestinyVersion,
  suggestionsContext: SuggestionsContext = {}
): SearchConfig {
  const suggestions = new Set<string>();
  const isFilters: Record<string, FilterDefinition> = {};
  const kvFilters: Record<string, FilterDefinition> = {};
  const allApplicableFilters: FilterDefinition[] = [];
  for (const filter of allFilters) {
    if (!filter.destinyVersion || filter.destinyVersion === destinyVersion) {
      for (const suggestion of generateSuggestionsForFilter(filter)) {
        suggestions.add(suggestion);
      }
      for (const suggestion of filter.suggestionsGenerator?.(suggestionsContext) ?? []) {
        suggestions.add(suggestion);
      }
      allApplicableFilters.push(filter);
      const filterKeywords = Array.isArray(filter.keywords) ? filter.keywords : [filter.keywords];
      const filterFormats = canonicalFilterFormats(filter.format);
      const hasSimple = filterFormats.some((f) => f === 'simple');
      const hasKv = filterFormats.some((f) => f !== 'simple');

      for (const keyword of filterKeywords) {
        if (hasSimple) {
          if ($DIM_FLAVOR === 'test' && isFilters[keyword]) {
            throw new Error(
              `Conflicting is:${keyword} filter -- only the last inserted filter will work.`
            );
          }
          isFilters[keyword] = filter;
        }
        if (hasKv) {
          if ($DIM_FLAVOR === 'test' && kvFilters[keyword]) {
            throw new Error(
              `Conflicting ${keyword}:value filter -- only the last inserted filter will work.`
            );
          }
          kvFilters[keyword] = filter;
        }
      }
    }
  }

  return {
    allFilters: allApplicableFilters,
    suggestions: Array.from(suggestions),
    isFilters,
    kvFilters,
  };
}
