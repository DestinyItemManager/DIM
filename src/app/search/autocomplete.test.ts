import { Search } from '@destinyitemmanager/dim-api-types';
import {
  autocompleteTermSuggestions,
  filterSortRecentSearches,
  makeFilterComplete,
} from './autocomplete';
import { buildSearchConfig } from './search-config';

/**
 * Given a string like "foo ba|r", find where the "|" is and remove it,
 * returning its index. This allows for readable test cases that depend on
 * cursor position. If the cursor should be at the end of the string, it can be
 * omitted entirely.
 */
function extractCaret(stringWithCaretPlaceholder: string): [caretIndex: number, query: string] {
  const caretIndex = stringWithCaretPlaceholder.indexOf('|');
  if (caretIndex == -1) {
    return [stringWithCaretPlaceholder.length, stringWithCaretPlaceholder];
  }
  return [caretIndex, stringWithCaretPlaceholder.replace('|', '')];
}

describe('autocompleteTermSuggestions', () => {
  const searchConfig = buildSearchConfig(2);
  const filterComplete = makeFilterComplete(searchConfig);

  const cases: string[] = [
    'is:haspower is:b',
    '(is:blue ju|n)',
    'is:bow is:v|oid',
    'season:>outl',
    'not(',
  ];

  test.each(cases)('autocomplete within query for {%s}', (queryWithCaret) => {
    const [caretIndex, query] = extractCaret(queryWithCaret);
    const candidates = autocompleteTermSuggestions(query, caretIndex, filterComplete, searchConfig);
    expect(candidates).toMatchSnapshot();
  });

  const multiWordCases: [query: string, mockCandidate: string][] = [
    ['arctic haz', 'arctic haze'],
    ['is:weapon arctic haz| -is:exotic', 'arctic haze'],
    ['name:"foo" arctic haz', 'arctic haze'],
    ["ager's sce", "ager's scepter"],
    ['the last word', 'the last word'],
    ['acd/0 fee', 'acd/0 feedback fence'],
    ['stat:rpm:200 first in, last', 'first in, last out'],
    ['two-tail', 'two-tailed fox'],
    ['(is:a or is:b) and (is:c or multi w|)', 'multi word'],
    ['arctic  haz', 'arctic haze'], // two spaces inbetween words
    ['"rare curio" arctic haz', 'arctic haze'],
    ['"rare curio" or arctic haz', 'arctic haze'],
    ['toil and trou', 'toil and trouble'], // todo: not handled due to the `and`
    ['rare curio or arctic haz', 'arctic haze'], // todo: parser result is unexpected here
  ];

  test.each(multiWordCases)(
    'autocomplete within multi-word query for {%s} with exact match',
    (queryWithCaret: string, mockCandidate: string) => {
      const [caretIndex, query] = extractCaret(queryWithCaret);
      const candidates = autocompleteTermSuggestions(
        query,
        caretIndex,
        // use mock candidates to simulate exact name-matches for multiword items
        () => [`name:"${mockCandidate}"`],
        searchConfig
      );
      expect(candidates).toMatchSnapshot();
    }
  );
});

describe('filterSortRecentSearches', () => {
  const recentSearches: Search[] = [
    { query: 'recent saved', usageCount: 1, saved: true, lastUsage: Date.now() },
    {
      query: 'yearold saved',
      usageCount: 1,
      saved: true,
      lastUsage: Date.now() - 365 * 24 * 60 * 60 * 1000,
    },
    {
      query: 'yearold unsaved',
      usageCount: 1,
      saved: false,
      lastUsage: Date.now() - 365 * 24 * 60 * 60 * 1000,
    },
    {
      query: 'yearold highuse',
      usageCount: 100,
      saved: false,
      lastUsage: Date.now() - 365 * 24 * 60 * 60 * 1000,
    },
    {
      query: 'dayold highuse',
      usageCount: 15,
      saved: false,
      lastUsage: Date.now() - 1 * 24 * 60 * 60 * 1000,
    },
    {
      query: 'dim api autosuggest',
      usageCount: 0,
      saved: false,
      lastUsage: 0,
    },
  ];

  for (let day = 0; day < 30; day++) {
    for (let usageCount = 1; usageCount < 10; usageCount++) {
      recentSearches.push({
        query: `${day} days old, ${usageCount} uses`,
        lastUsage: Date.now() - day * 24 * 60 * 60 * 1000,
        usageCount,
        saved: false,
      });
    }
  }

  const cases = [[''], ['high']];

  test.each(cases)('filter/sort recent searches for query |%s|', (query) => {
    const candidates = filterSortRecentSearches(query, recentSearches);
    expect(candidates.map((c) => c.query.fullText)).toMatchSnapshot();
  });

  const savedSearches: Search[] = [
    {
      query: 'is:patternunlocked -is:crafted',
      usageCount: 1,
      saved: true,
      lastUsage: Date.now(),
    },
    {
      query: '/* random-roll craftable guns */ is:patternunlocked -is:crafted',
      usageCount: 1,
      saved: true,
      lastUsage: Date.now() - 24 * 60 * 60 * 1000,
    },
  ];
  const highlightCases: string[] = ['', 'craft', 'craftable', 'crafted'];
  test.each(highlightCases)('check saved search highlighting for query |%s|', (query: string) => {
    const candidates = filterSortRecentSearches(query, savedSearches);
    expect(candidates).toMatchSnapshot();
  });
});

describe('filterComplete', () => {
  const searchConfig = buildSearchConfig(2);
  const filterComplete = makeFilterComplete(searchConfig);

  const terms = [['is:b'], ['jun'], ['sni'], ['stat:mob'], ['stat'], ['stat:'], ['ote']];

  test.each(terms)('autocomplete terms for |%s|', (term) => {
    const candidates = filterComplete(term);
    expect(candidates).toMatchSnapshot();
  });
});
