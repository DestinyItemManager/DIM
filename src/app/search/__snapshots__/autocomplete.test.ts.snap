// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`autocompleteTermSuggestions autocomplete within query for |(is:blue jun)| with caret at position 11 1`] = `
Array [
  Object {
    "helpText": undefined,
    "highlightRange": Array [
      9,
      17,
    ],
    "query": "(is:blue tag:junk)",
    "type": 3,
  },
]
`;

exports[`autocompleteTermSuggestions autocomplete within query for |is:bow is:void| with caret at position 11 1`] = `
Array [
  Object {
    "helpText": undefined,
    "highlightRange": Array [
      7,
      14,
    ],
    "query": "is:bow is:void",
    "type": 3,
  },
]
`;

exports[`autocompleteTermSuggestions autocomplete within query for |is:haspower is:b| with caret at position 16 1`] = `
Array [
  Object {
    "helpText": undefined,
    "highlightRange": Array [
      12,
      18,
    ],
    "query": "is:haspower is:bow",
    "type": 3,
  },
  Object {
    "helpText": undefined,
    "highlightRange": Array [
      12,
      19,
    ],
    "query": "is:haspower is:blue",
    "type": 3,
  },
]
`;

exports[`filterComplete autocomplete terms for |is:b| 1`] = `
Array [
  "is:bow",
  "is:blue",
]
`;

exports[`filterComplete autocomplete terms for |jun| 1`] = `
Array [
  "tag:junk",
]
`;

exports[`filterComplete autocomplete terms for |ote| 1`] = `
Array [
  "is:hasnotes",
  "notes:",
]
`;

exports[`filterComplete autocomplete terms for |sni| 1`] = `
Array [
  "is:sniperrifle",
]
`;

exports[`filterComplete autocomplete terms for |stat:| 1`] = `
Array [
  "stat:",
  "stat:rpm:",
  "stat:rof:",
  "stat:charge:",
  "stat:impact:",
  "stat:handling:",
  "stat:range:",
  "stat:stability:",
  "stat:reload:",
  "stat:magazine:",
  "stat:aimassist:",
  "stat:equipspeed:",
  "stat:velocity:",
  "stat:blastradius:",
  "stat:recoildirection:",
  "stat:drawtime:",
  "stat:zoom:",
  "stat:inventorysize:",
  "stat:swingspeed:",
  "stat:guardefficiency:",
  "stat:guardresistance:",
  "stat:chargerate:",
  "stat:guardendurance:",
  "stat:ammocapacity:",
  "stat:mobility:",
  "stat:resilience:",
  "stat:recovery:",
  "stat:discipline:",
  "stat:intellect:",
  "stat:strength:",
  "stat:total:",
  "stat:custom:",
  "stat:any:",
  "stat:rpm:<",
  "stat:rof:<",
  "stat:charge:<",
  "stat:impact:<",
  "stat:handling:<",
  "stat:range:<",
  "stat:stability:<",
  "stat:reload:<",
  "stat:magazine:<",
  "stat:aimassist:<",
  "stat:equipspeed:<",
  "stat:velocity:<",
  "stat:blastradius:<",
  "stat:recoildirection:<",
  "stat:drawtime:<",
  "stat:zoom:<",
  "stat:inventorysize:<",
  "stat:swingspeed:<",
  "stat:guardefficiency:<",
  "stat:guardresistance:<",
  "stat:chargerate:<",
  "stat:guardendurance:<",
  "stat:ammocapacity:<",
  "stat:mobility:<",
  "stat:resilience:<",
  "stat:recovery:<",
  "stat:discipline:<",
  "stat:intellect:<",
  "stat:strength:<",
  "stat:total:<",
  "stat:custom:<",
  "stat:any:<",
  "stat:rpm:>",
  "stat:rof:>",
  "stat:charge:>",
  "stat:impact:>",
  "stat:handling:>",
  "stat:range:>",
  "stat:stability:>",
  "stat:reload:>",
  "stat:magazine:>",
  "stat:aimassist:>",
  "stat:equipspeed:>",
  "stat:velocity:>",
  "stat:blastradius:>",
  "stat:recoildirection:>",
  "stat:drawtime:>",
  "stat:zoom:>",
  "stat:inventorysize:>",
  "stat:swingspeed:>",
  "stat:guardefficiency:>",
  "stat:guardresistance:>",
  "stat:chargerate:>",
  "stat:guardendurance:>",
  "stat:ammocapacity:>",
  "stat:mobility:>",
  "stat:resilience:>",
  "stat:recovery:>",
  "stat:discipline:>",
  "stat:intellect:>",
  "stat:strength:>",
  "stat:total:>",
  "stat:custom:>",
  "stat:any:>",
  "stat:rpm:<=",
  "stat:rof:<=",
  "stat:charge:<=",
  "stat:impact:<=",
  "stat:handling:<=",
  "stat:range:<=",
  "stat:stability:<=",
  "stat:reload:<=",
  "stat:magazine:<=",
  "stat:aimassist:<=",
  "stat:equipspeed:<=",
  "stat:velocity:<=",
  "stat:blastradius:<=",
  "stat:recoildirection:<=",
  "stat:drawtime:<=",
  "stat:zoom:<=",
  "stat:inventorysize:<=",
  "stat:swingspeed:<=",
  "stat:guardefficiency:<=",
  "stat:guardresistance:<=",
  "stat:chargerate:<=",
  "stat:guardendurance:<=",
  "stat:ammocapacity:<=",
  "stat:mobility:<=",
  "stat:resilience:<=",
  "stat:recovery:<=",
  "stat:discipline:<=",
  "stat:intellect:<=",
  "stat:strength:<=",
  "stat:total:<=",
  "stat:custom:<=",
  "stat:any:<=",
  "stat:rpm:>=",
  "stat:rof:>=",
  "stat:charge:>=",
  "stat:impact:>=",
  "stat:handling:>=",
  "stat:range:>=",
  "stat:stability:>=",
  "stat:reload:>=",
  "stat:magazine:>=",
  "stat:aimassist:>=",
  "stat:equipspeed:>=",
  "stat:velocity:>=",
  "stat:blastradius:>=",
  "stat:recoildirection:>=",
  "stat:drawtime:>=",
  "stat:zoom:>=",
  "stat:inventorysize:>=",
  "stat:swingspeed:>=",
  "stat:guardefficiency:>=",
  "stat:guardresistance:>=",
  "stat:chargerate:>=",
  "stat:guardendurance:>=",
  "stat:ammocapacity:>=",
  "stat:mobility:>=",
  "stat:resilience:>=",
  "stat:recovery:>=",
  "stat:discipline:>=",
  "stat:intellect:>=",
  "stat:strength:>=",
  "stat:total:>=",
  "stat:custom:>=",
  "stat:any:>=",
]
`;

exports[`filterComplete autocomplete terms for |stat:mob| 1`] = `
Array [
  "stat:mobility:",
]
`;

exports[`filterComplete autocomplete terms for |stat| 1`] = `
Array [
  "stat:",
  "basestat:",
  "maxstatloadout:",
  "maxstatvalue:",
  "maxbasestatvalue:",
  "stat:rpm:",
  "stat:rof:",
  "stat:charge:",
  "stat:impact:",
  "stat:handling:",
  "stat:range:",
  "stat:stability:",
  "stat:reload:",
  "stat:magazine:",
  "stat:aimassist:",
  "stat:equipspeed:",
  "stat:velocity:",
  "stat:blastradius:",
  "stat:recoildirection:",
  "stat:drawtime:",
  "stat:zoom:",
  "stat:inventorysize:",
  "stat:swingspeed:",
  "stat:guardefficiency:",
  "stat:guardresistance:",
  "stat:chargerate:",
  "stat:guardendurance:",
  "stat:ammocapacity:",
  "stat:mobility:",
  "stat:resilience:",
  "stat:recovery:",
  "stat:discipline:",
  "stat:intellect:",
  "stat:strength:",
  "stat:total:",
  "stat:custom:",
  "stat:any:",
  "basestat:mobility:",
  "basestat:resilience:",
  "basestat:recovery:",
  "basestat:discipline:",
  "basestat:intellect:",
  "basestat:strength:",
  "basestat:total:",
  "basestat:custom:",
  "basestat:any:",
  "maxstatloadout:mobility",
  "maxstatloadout:resilience",
  "maxstatloadout:recovery",
  "maxstatloadout:discipline",
  "maxstatloadout:intellect",
  "maxstatloadout:strength",
  "maxstatloadout:total",
  "maxstatloadout:custom",
  "maxstatloadout:any",
  "maxstatvalue:mobility",
  "maxstatvalue:resilience",
  "maxstatvalue:recovery",
  "maxstatvalue:discipline",
  "maxstatvalue:intellect",
  "maxstatvalue:strength",
  "maxstatvalue:total",
  "maxstatvalue:custom",
  "maxstatvalue:any",
  "maxbasestatvalue:mobility",
  "maxbasestatvalue:resilience",
  "maxbasestatvalue:recovery",
  "maxbasestatvalue:discipline",
  "maxbasestatvalue:intellect",
  "maxbasestatvalue:strength",
  "maxbasestatvalue:total",
  "maxbasestatvalue:custom",
  "maxbasestatvalue:any",
]
`;

exports[`filterSortRecentSearches filter/sort recent searches for query || 1`] = `
Array [
  "recent saved",
  "yearold saved",
  "0 days old, 9 uses",
  "0 days old, 8 uses",
  "dayold highuse",
  "0 days old, 7 uses",
  "1 days old, 9 uses",
  "1 days old, 8 uses",
  "0 days old, 6 uses",
  "1 days old, 7 uses",
  "2 days old, 9 uses",
  "2 days old, 8 uses",
  "1 days old, 6 uses",
  "2 days old, 7 uses",
  "3 days old, 9 uses",
  "0 days old, 5 uses",
  "3 days old, 8 uses",
  "2 days old, 6 uses",
  "3 days old, 7 uses",
  "4 days old, 9 uses",
  "1 days old, 5 uses",
  "4 days old, 8 uses",
  "3 days old, 6 uses",
  "4 days old, 7 uses",
  "5 days old, 9 uses",
  "2 days old, 5 uses",
  "5 days old, 8 uses",
  "4 days old, 6 uses",
  "0 days old, 4 uses",
  "5 days old, 7 uses",
  "6 days old, 9 uses",
  "3 days old, 5 uses",
  "6 days old, 8 uses",
  "5 days old, 6 uses",
  "1 days old, 4 uses",
  "6 days old, 7 uses",
  "7 days old, 9 uses",
  "4 days old, 5 uses",
  "7 days old, 8 uses",
  "6 days old, 6 uses",
  "2 days old, 4 uses",
  "7 days old, 7 uses",
  "8 days old, 9 uses",
  "5 days old, 5 uses",
  "8 days old, 8 uses",
  "7 days old, 6 uses",
  "3 days old, 4 uses",
  "8 days old, 7 uses",
  "9 days old, 9 uses",
  "6 days old, 5 uses",
  "9 days old, 8 uses",
  "8 days old, 6 uses",
  "4 days old, 4 uses",
  "9 days old, 7 uses",
  "10 days old, 9 uses",
  "0 days old, 3 uses",
  "7 days old, 5 uses",
  "10 days old, 8 uses",
  "9 days old, 6 uses",
  "5 days old, 4 uses",
  "10 days old, 7 uses",
  "11 days old, 9 uses",
  "1 days old, 3 uses",
  "8 days old, 5 uses",
  "11 days old, 8 uses",
  "10 days old, 6 uses",
  "6 days old, 4 uses",
  "11 days old, 7 uses",
  "12 days old, 9 uses",
  "2 days old, 3 uses",
  "9 days old, 5 uses",
  "12 days old, 8 uses",
  "11 days old, 6 uses",
  "7 days old, 4 uses",
  "12 days old, 7 uses",
  "13 days old, 9 uses",
  "3 days old, 3 uses",
  "10 days old, 5 uses",
  "13 days old, 8 uses",
  "12 days old, 6 uses",
  "8 days old, 4 uses",
  "13 days old, 7 uses",
  "14 days old, 9 uses",
  "4 days old, 3 uses",
  "11 days old, 5 uses",
  "14 days old, 8 uses",
  "13 days old, 6 uses",
  "9 days old, 4 uses",
  "14 days old, 7 uses",
  "15 days old, 9 uses",
  "5 days old, 3 uses",
  "12 days old, 5 uses",
  "15 days old, 8 uses",
  "14 days old, 6 uses",
  "10 days old, 4 uses",
  "15 days old, 7 uses",
  "16 days old, 9 uses",
  "6 days old, 3 uses",
  "13 days old, 5 uses",
  "16 days old, 8 uses",
  "15 days old, 6 uses",
  "11 days old, 4 uses",
  "16 days old, 7 uses",
  "17 days old, 9 uses",
  "7 days old, 3 uses",
  "14 days old, 5 uses",
  "17 days old, 8 uses",
  "0 days old, 2 uses",
  "16 days old, 6 uses",
  "12 days old, 4 uses",
  "17 days old, 7 uses",
  "18 days old, 9 uses",
  "8 days old, 3 uses",
  "15 days old, 5 uses",
  "18 days old, 8 uses",
  "1 days old, 2 uses",
  "17 days old, 6 uses",
  "13 days old, 4 uses",
  "18 days old, 7 uses",
  "19 days old, 9 uses",
  "9 days old, 3 uses",
  "16 days old, 5 uses",
  "19 days old, 8 uses",
  "2 days old, 2 uses",
  "18 days old, 6 uses",
  "14 days old, 4 uses",
  "19 days old, 7 uses",
  "20 days old, 9 uses",
  "10 days old, 3 uses",
  "17 days old, 5 uses",
  "20 days old, 8 uses",
  "3 days old, 2 uses",
  "19 days old, 6 uses",
  "15 days old, 4 uses",
  "20 days old, 7 uses",
  "21 days old, 9 uses",
  "11 days old, 3 uses",
  "18 days old, 5 uses",
  "21 days old, 8 uses",
  "4 days old, 2 uses",
  "20 days old, 6 uses",
  "16 days old, 4 uses",
  "21 days old, 7 uses",
  "22 days old, 9 uses",
  "12 days old, 3 uses",
  "19 days old, 5 uses",
  "22 days old, 8 uses",
  "5 days old, 2 uses",
  "21 days old, 6 uses",
  "17 days old, 4 uses",
  "22 days old, 7 uses",
  "23 days old, 9 uses",
  "13 days old, 3 uses",
  "20 days old, 5 uses",
  "23 days old, 8 uses",
  "6 days old, 2 uses",
  "22 days old, 6 uses",
  "18 days old, 4 uses",
  "23 days old, 7 uses",
  "24 days old, 9 uses",
  "14 days old, 3 uses",
  "21 days old, 5 uses",
  "24 days old, 8 uses",
  "7 days old, 2 uses",
  "23 days old, 6 uses",
  "19 days old, 4 uses",
  "24 days old, 7 uses",
  "25 days old, 9 uses",
  "15 days old, 3 uses",
  "22 days old, 5 uses",
  "25 days old, 8 uses",
  "8 days old, 2 uses",
  "24 days old, 6 uses",
  "20 days old, 4 uses",
  "25 days old, 7 uses",
  "26 days old, 9 uses",
  "16 days old, 3 uses",
  "23 days old, 5 uses",
  "26 days old, 8 uses",
  "9 days old, 2 uses",
  "25 days old, 6 uses",
  "21 days old, 4 uses",
  "26 days old, 7 uses",
  "27 days old, 9 uses",
  "17 days old, 3 uses",
  "24 days old, 5 uses",
  "27 days old, 8 uses",
  "10 days old, 2 uses",
  "26 days old, 6 uses",
  "22 days old, 4 uses",
  "27 days old, 7 uses",
  "28 days old, 9 uses",
  "18 days old, 3 uses",
  "25 days old, 5 uses",
  "28 days old, 8 uses",
  "11 days old, 2 uses",
  "27 days old, 6 uses",
  "23 days old, 4 uses",
  "28 days old, 7 uses",
  "29 days old, 9 uses",
  "19 days old, 3 uses",
  "26 days old, 5 uses",
  "29 days old, 8 uses",
  "12 days old, 2 uses",
  "28 days old, 6 uses",
  "24 days old, 4 uses",
  "29 days old, 7 uses",
  "20 days old, 3 uses",
  "27 days old, 5 uses",
  "13 days old, 2 uses",
  "29 days old, 6 uses",
  "25 days old, 4 uses",
  "21 days old, 3 uses",
  "28 days old, 5 uses",
  "14 days old, 2 uses",
  "26 days old, 4 uses",
  "22 days old, 3 uses",
  "29 days old, 5 uses",
  "15 days old, 2 uses",
  "27 days old, 4 uses",
  "0 days old, 1 uses",
  "23 days old, 3 uses",
  "16 days old, 2 uses",
  "28 days old, 4 uses",
  "1 days old, 1 uses",
  "24 days old, 3 uses",
  "17 days old, 2 uses",
  "29 days old, 4 uses",
  "2 days old, 1 uses",
  "25 days old, 3 uses",
  "18 days old, 2 uses",
  "3 days old, 1 uses",
  "26 days old, 3 uses",
  "19 days old, 2 uses",
  "4 days old, 1 uses",
  "27 days old, 3 uses",
  "20 days old, 2 uses",
  "5 days old, 1 uses",
  "28 days old, 3 uses",
  "21 days old, 2 uses",
  "6 days old, 1 uses",
  "29 days old, 3 uses",
  "22 days old, 2 uses",
  "7 days old, 1 uses",
  "23 days old, 2 uses",
  "8 days old, 1 uses",
  "24 days old, 2 uses",
  "9 days old, 1 uses",
  "25 days old, 2 uses",
  "10 days old, 1 uses",
  "26 days old, 2 uses",
  "11 days old, 1 uses",
  "27 days old, 2 uses",
  "12 days old, 1 uses",
  "28 days old, 2 uses",
  "13 days old, 1 uses",
  "29 days old, 2 uses",
  "14 days old, 1 uses",
  "15 days old, 1 uses",
  "16 days old, 1 uses",
  "17 days old, 1 uses",
  "18 days old, 1 uses",
  "19 days old, 1 uses",
  "20 days old, 1 uses",
  "21 days old, 1 uses",
  "22 days old, 1 uses",
  "23 days old, 1 uses",
  "24 days old, 1 uses",
  "25 days old, 1 uses",
  "26 days old, 1 uses",
  "27 days old, 1 uses",
  "28 days old, 1 uses",
  "29 days old, 1 uses",
  "yearold highuse",
  "yearold unsaved",
  "dim api autosuggest",
]
`;

exports[`filterSortRecentSearches filter/sort recent searches for query |high| 1`] = `
Array [
  "dayold highuse",
  "yearold highuse",
]
`;
